### PR TITLE
feat(advanced-search): Complete Advanced Search: backend changes & tying UI together

### PR DESCRIPTION
**Pull request by gabe-lyons**
This pr accomplishes a few things:

- Adds advanced search components introduced in previous PRs into the Search + Embedded Search + Impact Analysis pages.
- Sends advanced search filters to the backend
- Makes a **breaking change** to the graphql api to remove `value` and replace with `values` for FacetFilterInput graphql param. This means we will only send collections of values back for each filter. If only one value is provided, the array will have one element.
- Adds backend logic to use `values` instead of `value` from facet filter input, handle different conditions, handle negation, and handle OR/AND union types.
- Adds tests for new backend search logic + integration tests in cypress to apply a few different advanced search filter concepts
- Adds documentation for using advanced filters.

## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: datahub-graphql-core/src/main/resources/search.graphql [hunk 1]
```diff
@@ -150,22 +165,17 @@ input FacetFilterInput {
   field: String!
 
   """
-  Value of the field to filter by (soon to be deprecated)
-  """
-  value: String!
-
```
Comment (gabe-lyons) [inline datahub-graphql-core/src/main/resources/search.graphql:156]: note: this is a breaking change (we are removing the `value` field from `FacetFilterInput`

File: datahub-web-react/src/app/search/utils/filtersToGraphqlParams.ts [hunk 2]
```diff
@@ -1,10 +0,0 @@
-import { FacetFilterInput } from '../../../types.generated';
```
Comment (gabe-lyons) [inline datahub-web-react/src/app/search/utils/filtersToGraphqlParams.ts:1]: this file has been unused for some time.

File: docs/how/search.md [hunk 3]
```diff
@@ -1,46 +1,83 @@
...
+   Datasets has 2 attributes that contains field description. fieldDescription comes from the SchemaMetadata aspect, while editedFieldDescriptions comes from the EditableSchemaMetadata aspect. EditableSchemaMetadata holds information that comes from UI edits, while SchemaMetadata holds data from ingestion of the dataset.
+
+5. _Find a dataset with the term **logical** in the dataset description_  
+   `editedDescription: *logical* OR description: *logical*` [Sample results](https://demo.datahubproject.io/search?page=1&query=editedDescription%3A%20%2Alogical%2A%20OR%20description%3A%20%2Alogical%2A)  
+   Similar to field descriptions, dataset descriptions can be found in 2 aspects, hence the need to search 2 attributes.
+
+6. _Find a dataset which reside in one of the browsing folders, for instance, the **hive** folder_  
+   `browsePaths: *hive*` [Sample results](https://demo.datahubproject.io/search?page=1&query=browsePaths%3A%20%2Ahive%2A)  
+   BrowsePath is stored as a complete string, for instance `/datasets/prod/hive/SampleKafkaDataset`, hence the need for wildcards on both ends of the term to return a result.
+
```
Comment (gabe-lyons) [inline docs/how/search.md:74]: ignore- this just was reformatted

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java [hunk 4]
```diff
@@ -463,17 +477,21 @@ private void addCriterionFiltersToAggregationMetadata(
       originalMetadata.add(buildAggregationMetadata(
           finalFacetField,
           _filtersToDisplayName.getOrDefault(finalFacetField, finalFacetField),
-          new LongMap(ImmutableMap.of(criterion.getValue(), 0L)),
-          new FilterValueArray(ImmutableList.of(createFilterValue(criterion.getValue(), 0L))))
+          new LongMap(criterion.getValues().stream().collect(Collectors.toMap(i -> i, i -> 0L))),
+          new FilterValueArray(criterion.getValues().stream().map(value -> createFilterValue(value, 0L)).collect(
+              Collectors.toList())))
```
Comment (gabe-lyons) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java:482]: updates for `values`

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java [hunk 5]
```diff
@@ -489,12 +507,4 @@ private AggregationMetadata buildAggregationMetadata(
...
   }
 
-  @Nullable
-  private String toFacetField(@Nonnull final String filterField) {
-    String trimmedField = filterField.replace(ESUtils.KEYWORD_SUFFIX, "");
-    if (_facetFields.contains(trimmedField)) {
-      return trimmedField;
-    }
-    return null;
-  }
```
Comment (gabe-lyons) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java:499]: moved to ESUtils

File: docs/how/search.md [hunk 6]
```diff
@@ -1,46 +1,83 @@
...
+Using the Advanced Filter view, you can apply more complex filters. To get there, click 'Advanced' in the top right of the filter panel:
+
+<p align="center">
+  <img width="70%"  src="https://raw.githubusercontent.com/datahub-project/static-assets/main/imgs/advanced_search/click_to_advanced_search_view.png"/>
+</p>
+
+#### Adding an Advanced Filter
+
+Currently, Advanced Filters support filtering by Column name, Container, Domain, Description (entity or column level), Tag & Glossary Term (entity or column level), Owner, Entity Type, Subtype, Environment and soft-deleted status.
```
Comment (chriscollins3456) [inline docs/how/search.md:17]: should the `&` between Tag & Glossary Term just be a comma?

File: metadata-io/src/main/java/com/linkedin/metadata/search/aggregator/AllEntitiesSearchAggregator.java [hunk 7]
```diff
@@ -57,6 +61,7 @@ public AllEntitiesSearchAggregator(
     _searchRanker = Objects.requireNonNull(searchRanker);
     _cachingEntitySearchService = Objects.requireNonNull(cachingEntitySearchService);
     _entityDocCountCache = new EntityDocCountCache(entityRegistry, entitySearchService);
+    _entityRegistry = entityRegistry;
```
Comment (chriscollins3456) [inline metadata-io/src/main/java/com/linkedin/metadata/search/aggregator/AllEntitiesSearchAggregator.java:64]: why add `entityRegistry` here if it's not being used? Same thing with the two imports above

File: metadata-io/src/main/java/com/linkedin/metadata/search/utils/ESUtils.java [hunk 8]
```diff
@@ -107,12 +130,42 @@ public static BoolQueryBuilder buildConjunctiveFilterQuery(@Nonnull ConjunctiveC
...
+  }
+  @Nonnull
+  public static QueryBuilder getQueryBuilderFromCriterionForSingleField(@Nonnull Criterion criterion) {
     final Condition condition = criterion.getCondition();
+    String fieldName = toFacetField(criterion.getField());
+
     if (condition == Condition.EQUAL) {
       // If values is set, use terms query to match one of the values
       if (!criterion.getValues().isEmpty()) {
+        if (BOOLEAN_FIELDS.contains(fieldName) && criterion.getValues().size() == 1) {
```
Comment (chriscollins3456) [inline metadata-io/src/main/java/com/linkedin/metadata/search/utils/ESUtils.java:163]: if this is a boolean field then values shouldn't be anything other than 1, right? keeping the check regardless isn't a bad idea, just wanted to check for myself.

**Review by chriscollins3456 — COMMENTED** [review]
nice! there's a good amount of code here and stuff i'm less familiar with, but I don't have anything major here by any means.

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/search/SearchResolver.java [hunk 9]
```diff
@@ -41,17 +41,17 @@ public CompletableFuture<SearchResults> get(DataFetchingEnvironment environment)
...
       } catch (Exception e) {
-        log.error("Failed to execute search: entity type {}, query {}, filters: {}, start: {}, count: {}",
-            input.getType(), input.getQuery(), input.getFilters(), start, count);
+        log.error("Failed to execute search: entity type {}, query {}, filters: {}, orFilters: {}, start: {}, count: {}",
+            input.getType(), input.getQuery(), input.getFilters(), input.getOrFilters(), start, count);
         throw new RuntimeException(
-            "Failed to execute search: " + String.format("entity type %s, query %s, filters: %s, start: %s, count: %s",
-                input.getType(), input.getQuery(), input.getFilters(), start, count), e);
+            "Failed to execute search: " + String.format("entity type %s, query %s, filters: %s, orFilters: %s, start: %s, count: %s",
+                input.getType(), input.getQuery(), input.getFilters(), input.getOrFilters(), start, count), e);
```
[the remark this answers is not in the document]
Comment (gabe-lyons) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/search/SearchResolver.java:54]: unfortunately yes. Otherwise, we'd need to make breaking changes to the graphql api which we know folks are already making programatic calls to.

File: datahub-web-react/src/app/entity/shared/components/styled/search/EmbeddedListSearchModal.tsx [hunk 10]
```diff
@@ -70,9 +73,11 @@ export const EmbeddedListSearchModal = ({
                     query={query}
                     filters={filters}
                     page={page}
+                    unionType={unionType}
                     onChangeQuery={onChangeQuery}
                     onChangeFilters={onChangeFilters}
                     onChangePage={onChangePage}
+                    onChangeUnionType={setUnionType}
```
[the remark this answers is not in the document]
Comment (gabe-lyons) [reply datahub-web-react/src/app/entity/shared/components/styled/search/EmbeddedListSearchModal.tsx:80]: @chriscollins3456  just to keep consistency with the other function names in EmbeddedListSearch

File: docs/how/search.md (hunk 6 above)
Comment (gabe-lyons) [reply docs/how/search.md:17]: yes good call

File: metadata-io/src/main/java/com/linkedin/metadata/search/utils/ESUtils.java (hunk 8 above)
Comment (gabe-lyons) [reply metadata-io/src/main/java/com/linkedin/metadata/search/utils/ESUtils.java:163]: yes correct!

**Review by chriscollins3456 — APPROVED** [review]
looks good to me!

File: metadata-io/src/test/java/com/linkedin/metadata/search/SearchServiceTest.java [hunk 11]
```diff
@@ -160,4 +167,203 @@ public void testSearchService() throws Exception {
...
+
+    SearchResult searchResult =
+        _searchService.searchAcrossEntities(ImmutableList.of(ENTITY_NAME), "test", filterWithCondition, null, 0, 10, null);
+
+    assertEquals(searchResult.getNumEntities().intValue(), 0);
+    clearCache();
+
+    Urn urn = new TestEntityUrn("test", "testUrn", "VALUE_1");
+    ObjectNode document = JsonNodeFactory.instance.objectNode();
+    document.set("urn", JsonNodeFactory.instance.textNode(urn.toString()));
```
Comment (aditya-radhakrishnan) [inline metadata-io/src/test/java/com/linkedin/metadata/search/SearchServiceTest.java:202]: Nit: can we pull out some of these constants that we reuse? Also keeps things more extensible if we add more tests

File: docs/how/search.md [hunk 12]
```diff
@@ -39,6 +39,41 @@ The filters sidebar sits on the left hand side of search results, and lets users
  <img width="70%"  src="https://raw.githubusercontent.com/datahub-project/static-assets/main/imgs/filters_highlighted.png" />
 </p>
 
+### Advanced Filters
```
Comment (aditya-radhakrishnan) [inline docs/how/search.md:42]: Thank you so much for updating these docs!

File: docs/how/search.md [hunk 13]
```diff
@@ -39,6 +39,41 @@ The filters sidebar sits on the left hand side of search results, and lets users
...
+
+Using the Advanced Filter view, you can apply more complex filters. To get there, click 'Advanced' in the top right of the filter panel:
+
+<p align="center">
+  <img width="70%"  src="https://raw.githubusercontent.com/datahub-project/static-assets/main/imgs/advanced_search/click_to_advanced_search_view.png"/>
+</p>
+
+#### Adding an Advanced Filter
+
+Currently, Advanced Filters support filtering by Column name, Container, Domain, Description (entity or column level), Tag (entity or column level), Glossary Term (entity or column level), Owner, Entity Type, Subtype, Environment and soft-deleted status.
```
Comment (aditya-radhakrishnan) [inline docs/how/search.md:52]: Nit: I would capitalize Column Name

File: smoke-test/tests/cypress/cypress/integration/search/search.js [hunk 14]
```diff
@@ -1,65 +1,195 @@
...
-  it('can hit all entities search, see some results (testing this any more is tricky because it is cached for now)', () => {
-    cy.login();
-    cy.visit('/');
-    cy.get('input[data-testid=search-input]').type('*{enter}');
-    cy.wait(5000);
-	  cy.contains('of 0 results').should('not.exist');
-	  cy.contains(/of [0-9]+ results/);
-  });
+describe("search", () => {
+  // it("can hit all entities search, see some results (testing this any more is tricky because it is cached for now)", () => {
```
Comment (aditya-radhakrishnan) [inline smoke-test/tests/cypress/cypress/integration/search/search.js:2]: are these meant to be commented out....?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/assertion/AssertionRunEventResolver.java [hunk 15]
```diff
@@ -102,13 +100,16 @@ private Filter buildFilter(@Nullable FilterInput filtersInput, @Nullable final S
     }
     List<FacetFilterInput> facetFilters = new ArrayList<>();
     if (status != null) {
-      facetFilters.add(new FacetFilterInput("status", status, ImmutableList.of(status), false, SearchCondition.EQUAL));
+      FacetFilterInput filter = new FacetFilterInput();
+      filter.setField("status");
```
Comment (aditya-radhakrishnan) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/assertion/AssertionRunEventResolver.java:104]: can we make this a constant?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java [hunk 16]
```diff
@@ -81,20 +83,73 @@ public static Map<String, String> buildFacetFilters(@Nullable List<FacetFilterIn
...
+            facetFilters.put(facetFilterInput.getField(), facetFilterInput.getValues().get(0));
         });
 
         return facetFilters;
     }
 
+    // In the case that user sends filters to be or-d together, we need to build a series of conjective criterion
+    // arrays, rather than just one for the AND case.
+    public static ConjunctiveCriterionArray buildConjunctiveCriterionArrayWithOr(
+        final List<Criterion> andCriterions,
```
Comment (aditya-radhakrishnan) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:95]: do we need an annotation for `andCriterions` too?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java [hunk 17]
```diff
@@ -81,20 +83,73 @@ public static Map<String, String> buildFacetFilters(@Nullable List<FacetFilterIn
...
+                );
+            }
+        ).collect(Collectors.toList()));
+    }
+
     @Nullable
-    public static Filter buildFilter(@Nullable List<FacetFilterInput> facetFilterInputs) {
-        if (facetFilterInputs == null || facetFilterInputs.isEmpty()) {
+    public static Filter buildFilter(@Nullable List<FacetFilterInput> andFilters, @Nullable List<FacetFilterInput> orFilters) {
+        if ((andFilters == null || andFilters.isEmpty()) && (orFilters == null || orFilters.isEmpty())) {
```
Comment (aditya-radhakrishnan) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:111]: can you separate these two conditions out into boolean variables? it's a little hard to read, and you can use the the same variables down below when building the andCriterions and orCriterions.

Maybe `hasAndFilters` and `hasOrFilters`?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java [hunk 18]
```diff
@@ -81,20 +83,73 @@ public static Map<String, String> buildFacetFilters(@Nullable List<FacetFilterIn
...
-            facetFilters.put(facetFilterInput.getField(), facetFilterInput.getValue());
+            facetFilters.put(facetFilterInput.getField(), facetFilterInput.getValues().get(0));
         });
 
         return facetFilters;
     }
 
+    // In the case that user sends filters to be or-d together, we need to build a series of conjective criterion
+    // arrays, rather than just one for the AND case.
+    public static ConjunctiveCriterionArray buildConjunctiveCriterionArrayWithOr(
```
Comment (aditya-radhakrishnan) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:94]: nit: annotation this `@Nonnull`

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java [hunk 19]
```diff
@@ -81,20 +83,73 @@ public static Map<String, String> buildFacetFilters(@Nullable List<FacetFilterIn
...
                 throw new ValidationException(String.format("Unrecognized facet with name %s provided", facetFilterInput.getField()));
             }
-            facetFilters.put(facetFilterInput.getField(), facetFilterInput.getValue());
+            facetFilters.put(facetFilterInput.getField(), facetFilterInput.getValues().get(0));
         });
 
         return facetFilters;
     }
 
+    // In the case that user sends filters to be or-d together, we need to build a series of conjective criterion
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:92]: nit: conjunctive criterion?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java [hunk 20]
```diff
@@ -81,20 +83,73 @@ public static Map<String, String> buildFacetFilters(@Nullable List<FacetFilterIn
             if (!validFacetFields.contains(facetFilterInput.getField())) {
                 throw new ValidationException(String.format("Unrecognized facet with name %s provided", facetFilterInput.getField()));
             }
-            facetFilters.put(facetFilterInput.getField(), facetFilterInput.getValue());
+            facetFilters.put(facetFilterInput.getField(), facetFilterInput.getValues().get(0));
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:86]: ^

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java (hunk 18 above)
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:94]: Just leaving this here -> The name of this function + the parameters it takes are not immediately making it clear to me what this function does

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java [hunk 21]
```diff
@@ -81,20 +83,73 @@ public static Map<String, String> buildFacetFilters(@Nullable List<FacetFilterIn
...
         return facetFilters;
     }
 
+    // In the case that user sends filters to be or-d together, we need to build a series of conjective criterion
+    // arrays, rather than just one for the AND case.
+    public static ConjunctiveCriterionArray buildConjunctiveCriterionArrayWithOr(
+        final List<Criterion> andCriterions,
+        @Nonnull List<FacetFilterInput> orFilters
+    ) {
+        return new ConjunctiveCriterionArray(orFilters.stream().map(orFilter -> {
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:98]: So we are combining the existing "and" filters with a set of OR filters? Why is this necessary?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java [hunk 22]
```diff
@@ -81,20 +83,73 @@ public static Map<String, String> buildFacetFilters(@Nullable List<FacetFilterIn
...
+                CriterionArray andCriterionsForOr = new CriterionArray(andCriterions);
+                andCriterionsForOr.add(criterionFromFilter(orFilter));
+
+                return new ConjunctiveCriterion().setAnd(
+                    andCriterionsForOr
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:103]: So here we are doing something like this...

If you pass a list of and criterion:

`[ field equals value, field2 equals value2 ]`

and some set of ORs

`[ field3 equals value3 ]`

then we will essentially build an OR of ANDs?

`OR [  [ field equals value] AND [ field3 equals value3 ],  [field2 equals value2 ] AND [ field3 equals value3 ]  ]`

Comment (jjoyce0510) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:103]: So I had to work through this in my head for some time. We should consider adding some comments or something to give an example !

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java [hunk 23]
```diff
@@ -81,20 +83,73 @@ public static Map<String, String> buildFacetFilters(@Nullable List<FacetFilterIn
...
-        if (facetFilterInputs == null || facetFilterInputs.isEmpty()) {
+    public static Filter buildFilter(@Nullable List<FacetFilterInput> andFilters, @Nullable List<FacetFilterInput> orFilters) {
+        if ((andFilters == null || andFilters.isEmpty()) && (orFilters == null || orFilters.isEmpty())) {
             return null;
         }
-        return new Filter().setOr(new ConjunctiveCriterionArray(new ConjunctiveCriterion().setAnd(new CriterionArray(facetFilterInputs.stream()
-            .map(filter -> new Criterion().setField(getFilterField(filter.getField())).setValue(filter.getValue()))
-            .collect(Collectors.toList())))));
+
+        final List<Criterion> andCriterions = andFilters != null && !andFilters.isEmpty()
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:115]: This function is a bit confusing now. I was imagining we'd just have a nested OR or ANDs coming in from the FacetFilterInput. It's kinda hard to understand that if you provide both AND and OR filters that we'll combine them (and how)

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/assertion/AssertionRunEventResolver.java (hunk 15 above)
Comment (jjoyce0510) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/assertion/AssertionRunEventResolver.java:104]: +1

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/search/SearchAcrossEntitiesResolver.java [hunk 24]
```diff
@@ -52,7 +52,7 @@ public CompletableFuture<SearchResults> get(DataFetchingEnvironment environment)
             "Executing search for multiple entities: entity types {}, query {}, filters: {}, start: {}, count: {}",
             input.getTypes(), input.getQuery(), input.getFilters(), start, count);
         return UrnSearchResultsMapper.map(_entityClient.searchAcrossEntities(entityNames, sanitizedQuery,
-            ResolverUtils.buildFilter(input.getFilters()), start, count, ResolverUtils.getAuthentication(environment)));
+            ResolverUtils.buildFilter(input.getFilters(), input.getOrFilters()), start, count, ResolverUtils.getAuthentication(environment)));
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/search/SearchAcrossEntitiesResolver.java:55]: A bit odd - input.getFilters vs input.getOrFilters -> are we saying we are just taking on this model semantics debt ?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/search/SearchResolver.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/search/SearchResolver.java:54]: I'd prefer to see us taking our GraphQL API and making it an OR of ANDs

File: datahub-graphql-core/src/main/resources/search.graphql [hunk 25]
```diff
@@ -65,9 +65,14 @@ input SearchInput {
...
   """
-  Facet filters to apply to search results
+  Facet filters to apply to search results. These will be 'AND'-ed together.
   """
   filters: [FacetFilterInput!]
+
+  """
+  Filters are by default 'AND'-ed together. This lets you provide a set of OR filters
+  """
+  orFilters: [FacetFilterInput!]
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/resources/search.graphql:75]: Could we use a nested model instead of reusing FacetFilterInput?

basically:


```
type OrFilter {
    and: [FacetFilterInput!] 
}
```

This is a standard way to define filters, it's [Disjunctive Normal Form ](https://en.wikipedia.org/wiki/Disjunctive_normal_form)

Comment (jjoyce0510) [reply datahub-graphql-core/src/main/resources/search.graphql:75]: The reason i call this out is that once we commit this code it's going to become very nontrivial to change this, for the same reason its hard to change the existing semantics around AND filters..

Comment (jjoyce0510) [reply datahub-graphql-core/src/main/resources/search.graphql:75]: And unsurprisingly, this is how we chose to model the Filter.pdl model already, because we were looking ahead at this. 

It's shaped as 


{
    or: [
       {
             and: [ { field: field, value: value } ]
       }
    ]
}

Comment (gabe-lyons) [reply datahub-graphql-core/src/main/resources/search.graphql]: updated!

File: datahub-graphql-core/src/main/resources/recommendation.graphql [hunk 26]
```diff
@@ -237,12 +237,22 @@ type Filter {
...
+
+  """
+  If the filter should or should not be matched
+  """
+  negated: Boolean
+
+  """
+  Condition for the values. How to If unset, assumed to be equality
+  """
+  condition: SearchCondition
```
[the remark this answers is not in the document]
Comment (gabe-lyons) [reply datahub-graphql-core/src/main/resources/recommendation.graphql]: renamed

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/search/SearchAcrossEntitiesResolver.java (hunk 24 above)
Comment (gabe-lyons) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/search/SearchAcrossEntitiesResolver.java:55]: resolved in my update

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java (hunk 23 above)
Comment (gabe-lyons) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java]: resolved in my update- clearer now!

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java (hunk 22 above)
Comment (gabe-lyons) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java]: resovled!

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java (hunk 18 above)
Comment (gabe-lyons) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:105]: resolved!

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java (hunk 19 above)
Comment (gabe-lyons) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java]: fixed!

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java (hunk 18 above)
Comment (gabe-lyons) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:105]: resolved

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java (hunk 17 above)
Comment (gabe-lyons) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java:119]: resolved

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java (hunk 16 above)
Comment (gabe-lyons) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java]: no longer relevant

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/assertion/AssertionRunEventResolver.java (hunk 15 above)
Comment (gabe-lyons) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/assertion/AssertionRunEventResolver.java:104]: we can do this in a followup

File: smoke-test/tests/cypress/cypress/integration/search/search.js (hunk 14 above)
Comment (gabe-lyons) [reply smoke-test/tests/cypress/cypress/integration/search/search.js]: this has been reverted already :)

File: docs/how/search.md (hunk 13 above)
Comment (gabe-lyons) [reply docs/how/search.md]: fixed

File: docs/how/search.md (hunk 12 above)
Comment (gabe-lyons) [reply docs/how/search.md:42]: of course :)

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (gabe-lyons) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/ResolverUtils.java]: resolved
